### PR TITLE
Make parsing of LLM response more robust

### LINKED ISSUE #122: [Bug] Incorrect parsing of output file
We are seeing instances where the LLM is giving us extra information to help us, things like suggestions to pom.xml updates or updates to properties files.

We have bugs in our output parsing when we are looking for the updated java file, and we miss the java contents and we pick up on the xml/properites snippets

Here are 2 examples:

- https://github.com/konveyor-ecosystem/coolstore/pull/13/files#diff-05b046131cf88fd11b4a01ae3c81614bac720ca6083f808bd1d1400d1ac09531
- https://github.com/konveyor-ecosystem/coolstore/pull/13/files#diff-b41339bc26d1203736b6b9960e76de30071c0319986283bfb2cc169517217208

Here is another datapoint of an example that yielded (from a notebook) a properties file and an xml file:
https://github.com/konveyor-ecosystem/kai/blob/main/notebooks/jpa/output/gpt-4-1106-preview/cmt/kai/quarkus/persistence-to-quarkus-00000/few_shot/result.txt

### PR DESCRIPTION
**Pull request by fabianvf**
Fixes #122 
- Add language guesser and use it to more intelligently select code blocks from response
- Don't count incomplete responses as successful
  - Retries when updated file content is not parsed
  - Only attempts to find code block content underneath the Updated File header in order to avoid grabbing extra info from reasoning
- Add support for returning LLM results directly on the server and saving it from the demo client

### CODE REVIEW

**Review by JonahSussman — APPROVED** [review]
Love when there's prior work for stuff like this +1

File: example/run_demo.py [hunk 1]
```diff
@@ -155,6 +156,18 @@ def write_to_disk(file_path, updated_file_contents):
         KAI_LOG.error(f"Failed to write reasoning @ {reasoning_path} with error: {e}")
         KAI_LOG.error(f"Contents: {updated_file_contents}")
         sys.exit(1)
+    if updated_file_contents.get("llm_results"):
+        llm_result_path = f"{intended_file_path}.llm_result"
+        KAI_LOG.info(f"Writing reasoning to {llm_result_path}")
```
Comment (jwmatthews) [inline example/run_demo.py:161]: Assume we want to change 'Writing reasoning...' to 'Writing llm_result...'

File: kai/pydantic_models.py [hunk 2]
```diff
@@ -1,22 +1,74 @@
...
+        updated_file = code_block_matches[0]
+        if len(code_block_matches) > 1:
+            KAI_LOG.debug(
+                f"Multiple codeblocks found, defaulting to first {code_block_matches}"
+            )
+        else:
+            KAI_LOG.debug(f"Found single codeblock \n{updated_file}")
+    else:
+        updated_file = ""
+        KAI_LOG.warn("No codeblocks detected in LLM response")
```
Comment (jwmatthews) [inline kai/pydantic_models.py:71]: I'm seeing this more than I think would be expected, I'm worried something is not correct.
```

WARNING - 2024-03-28 13:52:30,093 - [pydantic_models.py:71 - parse_file_solution_content() ] - No codeblocks detected in LLM response
WARNING - 2024-03-28 13:52:30,094 - [server.py:534 - get_incident_solutions_for_file() ] - Request to model failed for batch 1/1 for src/main/java/com/redhat/coolstore/utils/StartupListener.java with exception, retrying in 10s


WARNING - 2024-03-28 13:53:31,175 - [pydantic_models.py:71 - parse_file_solution_content() ] - No codeblocks detected in LLM response
WARNING - 2024-03-28 13:53:31,177 - [server.py:534 - get_incident_solutions_for_file() ] - Request to model failed for batch 1/1 for src/main/java/com/redhat/coolstore/rest/CartEndpoint.java with exception, retrying in 10s


WARNING - 2024-03-28 13:56:04,740 - [pydantic_models.py:71 - parse_file_solution_content() ] - No codeblocks detected in LLM response
WARNING - 2024-03-28 13:56:04,742 - [server.py:534 - get_incident_solutions_for_file() ] - Request to model failed for batch 1/1 for src/main/java/com/redhat/coolstore/rest/CartEndpoint.java with exception, retrying in 10s
Error in LLM Response: The LLM did not provide an updated file for src/main/java/com/redhat/coolstore/rest/CartEndpoint.java
WARNING - 2024-03-28 13:58:52,980 - [pydantic_models.py:71 - parse_file_solution_content() ] - No codeblocks detected in LLM response
WARNING - 2024-03-28 13:58:52,983 - [server.py:534 - get_incident_solutions_for_file() ] - Request to model failed for batch 1/1 for src/main/java/com/redhat/coolstore/rest/CartEndpoint.java with exception, retrying in 10s
Error in LLM Response: The LLM did not provide an updated file for src/main/java/com/redhat/coolstore/rest/CartEndpoint.java
WARNING - 2024-03-28 14:01:47,964 - [pydantic_models.py:71 - parse_file_solution_content() ] - No codeblocks detected in LLM response
WARNING - 2024-03-28 14:01:47,965 - [server.py:534 - get_incident_solutions_for_file() ] - Request to model failed for batch 1/1 for src/main/java/com/redhat/coolstore/rest/CartEndpoint.java with exception, retrying in 10s


```

**Review by jwmatthews — CHANGES_REQUESTED** [review]
I worry that we need to spend more time on this problem, I think we need to be confident we have a solid way to parse the updated file out, it's the critical element of the call.

Let's dig into the below behavior and understand why we see it frequently.

[pydantic_models.py:71 - parse_file_solution_content() ] - No codeblocks detected in LLM response

File: kai/pydantic_models.py (hunk 2 above)
Comment (fabianvf) [reply kai/pydantic_models.py:71]: it's because mixtral is not providing a complete response

**Review by jwmatthews — APPROVED** [review]
LGTM 

My understanding is that the errors we see with `[pydantic_models.py:71 - parse_file_solution_content() ] - No codeblocks detected in LLM response` are _not_ related to this PR.   This PR is performing as intended, there is a deeper issue with mixtral, perhaps with our prompt that is causing the problems.  We will track that it in issue #126

**Pull request merged**